common-app-lib/src/cmr/common_app/api/request_logger.clj

```diff
@@ -0,0 +1,169 @@
+        clean-query (if (nil? token)
+                      raw-query
+                      (assoc raw-query "token" (util/scrub-token token)))
+        encoded-query (codec/form-encode clean-query)]
+    ;; build: <protocol>://<server>:<port><path><question-mark><params>
+    (format "%s://%s:%s%s%s%s"
+            (when request-scheme (name request-scheme))
+            (or (:server-name request) nil)
+            (or (:server-port request) nil)
+            (or (:uri request) nil)
+            (if (empty? encoded-query) "" "?") ;; don't show if no query
+            encoded-query)))
+
+(defn- scrub-token-from-map
+  "Scrub out a token keyed at field-name from a map"
+  ([data] (scrub-token-from-map data :token))
+  ([data field-name]
+   (let [token (get data field-name)]
+     (if token
+       (assoc data field-name (util/scrub-token token))
+       data))))
+
+(defn- dump-param
+  "Dump out a form post parameter or url parameter map from inside a given
+   request object taking care to mask tokens and to limit the number of actual
+   values that will be logged. Defaults (2 parameter call) to 128 parameters."
+  ([data field] (dump-param data field 128))
+  ([data field limit]
+   (-> data
+       (get field {})
+       ;; check for and scrub out token values to make safe for logging
+       (scrub-token-from-map)
+       (scrub-token-from-map "token")
+       ;; Parameters in forms could be really large, don't log everything
+       (as-> item (apply dissoc item (drop limit (keys item)))))))
+
+;; *****************************************************************************
+;; Ring Middleware Handlers
+
+(defn add-body-hashes
+  "Calculate the MD5 and SHA1 hash for the body of a response and include those
+   results in the response as HTTP headers. Headers are called Content-MD5 and
+   Content-SHA1. Content-MD5 is mentioned on the Wikipedia page:
+   https://en.wikipedia.org/wiki/List_of_HTTP_header_fields#Standard_request_fields
+   so it is presumed to be in common use."
+  [handler]
+  (fn [request]
+    (if-not (config/add-hash-headers)
+      (handler request)
+      (let [response (handler request)
+            ;; This is run after all responses
+            text (str (:body response))
+            body-md5 (digest/md5 text)
+            body-sha1 (digest/sha-1 text)
+            updated-response (-> response
+                                 (assoc-in [:headers "Content-MD5"] body-md5)
+                                 (assoc-in [:headers "Content-SHA1"] body-sha1))]
+        updated-response))))
+
+;; Action-logger should provide the same info as a standard NCSA Log
+;; ; 127.0.0.1 - - [2023-12-27 19:04:01.676] "GET /collections?keyword=any HTTP/1.1" 200 112 "-" "curl/8.1.2" 296
+
+;; Fields are as follows:
+;; ;
+;; IP Address
+;; login (ignored)
+;; -
+;; login (ignored)
+;; -
+;; [date-time]
+;; "
+;; HTTP VERB
+;; Relative URL
+;; "
+;; Status code
+;; response size
+;; "-"
+;; agent
+;; time
+
+(defn action-logger
```

Comment:
Can we name funcs to be more self-documenting?
Something like 'log-ring-request'